src/input.rs

```diff
@@ -410,14 +440,79 @@ impl<'a, A: ActionContext + 'a> Processor<'a, A> {
 
 #[cfg(test)]
 mod tests {
-    use glutin::{mods, VirtualKeyCode};
+    use glutin::{mods, VirtualKeyCode, Event, ElementState, MouseButton};
 
-    use term::mode;
+    use term::{SizeInfo, Term, mode};
+    use event::{self, Mouse, ClickState, ActionContext};
+    use event_loop::{self};
+    use config::{Config};
+    use selection::Selection;
+    use std::sync::{mpsc, Arc};
+    use sync::FairMutex;
+    use mio::{self};
 
-    use super::{Action, Binding};
+    use super::{Action, Binding, Processor};
 
     const KEY: VirtualKeyCode = VirtualKeyCode::Key0;
 
+    macro_rules! test_clickstate {
+        {
+            name: $name:ident,
+            initial_state: $initial_state:expr,
+            input: $input:expr,
+            end_state: $end_state:pat
+        } => {
+            #[test]
+            fn $name() {
+                let config = Config::default();
+                let size = SizeInfo {
+                    width: 21.0,
+                    height: 51.0,
+                    cell_width: 3.0,
+                    cell_height: 3.0,
+                };
+
+                let terminal = Term::new(&config, size);
+                let terminal = Arc::new(FairMutex::new(terminal));
+
+                let (tx, ..) = mpsc::channel();
+                let (sender, ..) = mio::channel::channel();
```

Comment:
You shouldn't need all of this setup anymore. If the point is just to test input processing, we should have an `ActionContext` implementation just for the tests.